File: _06_Heap_sort.java
import java.util.Scanner;

public class _06_Heap_sort {
    //heapsort:
    public static void heapSort(int[] array, int n){
        //Build maxheap:
        for(int i= n/2; i>=1; i--){
            maxHeapify(array,n,i);
        }
        //Deleting : 
        for(int i = n; i >= 2; i--){
            int temp = array[1];
            array[1] = array[i];
            array[i] = temp;

            maxHeapify(array, i-1, 1);
        }
    }
    //maxHeapify:
    public static void maxHeapify(int[] array, int n, int i){
        int largest = i;
        int l = 2*i;
        int r = 2*i + 1;

        if(l<=n && array[largest]<array[l]){
            largest = l;
        }
        if(r<=n && array[largest]<array[r]){
            largest = r;
        }
        if(largest!=i){
            array[largest] = array[largest] + array[i];
            array[i] = array[largest] - array[i];
            array[largest] = array[largest] - array[i];

            maxHeapify(array, n, largest);
        }

    }
//Main program : 
    public static void main(String[] args) {
        int[] array = input();
        int n = array.length-1;

        heapSort(array,n);
        System.out.println("\nAfter Heap Sorting : ");
        printArray(array,n);
    }
//Input : 
    public static int[] input(){
        int n;
        System.out.println("Enter the size of the array : ");
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        int[] array = new int[n+1];
        System.out.println("\nEnter the array : ");
        for(int i = 1; i <= n; i++){
            array[i] = scanner.nextInt();
        }
        scanner.close();
        return array;
    }
    //Printing : 
    public static void printArray(int[] array, int n){
        for(int i = 1; i <= n; i++){
            System.out.print(array[i]+" ");
        }System.out.println();
    }
}

File: _06_Heap_sort.cpp
#include<iostream>
using namespace std;

void printArray(int array[], int n);
void maxHeapify(int array[], int n, int i);
//HeapSort:
void heapSort(int array[], int n){
    //building : 
    for (int i = n/2; i >= 1; i--)
    {
        maxHeapify(array,n,i);
    }
    //Deleting :
    for (int i = n; i >= 2; i--)
    {
        swap(array[1],array[i]);
        maxHeapify(array,i-1,1);
    }
}

//MaxHeapify:
void maxHeapify(int array[], int n, int i){
    int largest = i;
    int l = 2*i;
    int r = 2*i + 1;
    
    if ( l <= n && array[largest]<array[l])
    {
        largest = l;
    }

    if ( r <= n && array[largest]<array[r])
    {
        largest = r;
    }
    
    if(largest!=i){
        swap(array[largest],array[i]);
        maxHeapify(array,n,largest);
    }
    
}

int main(){
    int n;
    cout << "Enter the size of the array : \n";
    cin >> n;
    int array[n+1];
    cout << "Enter the array : \n";
    for(int i = 1; i <= n; i++)
    {
        cin >> array[i];
    }

    heapSort(array, n);

    cout << "\nAfter Heap Sorting :\n";
    printArray(array,n);
}

//printing : 
void printArray(int array[], int n){
    for (int i = 1; i <= n; i++)
    {
        cout << array[i]<< " ";
    }cout << endl;
}